neuralcompression/models/_unet_discriminator.py

```diff
@@ -0,0 +1,337 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import math
+from functools import partial
+from typing import Callable, List, Optional, Union
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch import Tensor
+from torch.nn.utils.parametrizations import spectral_norm
+
+from ._discriminator import Discriminator
+
+
+def _verify_norm_type(norm_type: str):
+    if norm_type not in ("batch", "spectral", "instance", "identity"):
+        raise ValueError(f"Unrecognized norm_type {norm_type}")
+
+
+def apply_conv_norm(module: nn.Module, norm_type: str = "identity") -> nn.Module:
+    _verify_norm_type(norm_type)
+    if norm_type == "spectral":
+        return spectral_norm(module)
+    else:
+        return module
+
+
+def build_norm_layer(chans: int, norm_type: str) -> nn.Module:
+    _verify_norm_type(norm_type)
+    if norm_type == "batch":
+        return nn.BatchNorm2d(chans)
+    elif norm_type == "instance":
+        return nn.InstanceNorm2d(chans)
+    else:
+        return nn.Identity()
+
+
+def _conv(
+    in_planes: int,
+    out_planes: int,
+    kernel_size: int = 3,
+    stride: int = 1,
+    padding: int = 1,
+    bias: bool = True,
+):
+    return nn.Conv2d(
+        in_planes,
+        out_planes,
+        kernel_size=kernel_size,
+        stride=stride,
+        padding=padding,
+        bias=bias,
+    )
+
+
+class _UnetLevel(nn.Module):
+    def __init__(
+        self,
+        in_planes: int,
+        down_planes: int,
+        child: nn.Module,
```

Comment:
Same comment as before: the `child` is an implicit data structure in disguise that could be replaced with List, nn.Sequential or nn.ModuleList

Reply:
Just a personal preference, I've usually liked this sort of tree structure because it discards intermediate variables as you progress back up the tree. Usually with the list implementation you keep everything around until the end. Saves a small amount of memory.